DirectProgramming/DPC++FPGA/Tutorials/Features/private_copies/src/private_copies.cpp

```diff
@@ -0,0 +1,211 @@
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+
+  // The performance of the kernel is measured in GFlops, based on:
+  // 1) the number of inting-point operations performed by the kernel.
+  //    This can be calculated easily for the simple example kernel.
+  // 2) the kernel execution time reported by SYCL event profiling.
+  std::cout << "Num private_copies " << NumCopies << " "
+            << "kernel time : " << kernel_time << " ms\n";
+  std::cout << "Throughput for kernel with private_copies " << NumCopies
+            << ": ";
+  std::cout << std::fixed << std::setprecision(3)
+            << ((double)(kTotalOps) / kernel_time) / 1e6f << " GFlops\n";
+}
+
+// Calculates the expected results. Used to verify that the kernel
+// is functionally correct.
+int GoldenResult(const IntArray &input_arr, int shift) {
+  int gr = 0;
+
+  for (size_t i = 0; i < kMaxIter; i++) {
+    int a[kSize];
+    int b[kSize];
+    int c[kSize];
+    for (size_t j = 0; j < kSize; j++) {
+      a[j] = input_arr[(i * 4 + j) % kSize] * shift;
+    }
+    for (size_t j = 0; j < kSize; j++)
+      b[j] = SimpleMath(a[j], a[kSize - j - 1]);
+    for (size_t j = 0; j < kSize; j++)
+      c[j] = SimpleMath(b[j], b[kSize - j - 1]);
+    for (size_t j = 0; j < kSize; j++) 
+      gr += c[j];
+  }
+
+  return gr;
+}
+
+int main() {
+  bool success = true;
+
+  IntArray a;
+  IntScalar R0, R1, R2, R3, R4, R5;
+
+  int shift = rand() % kMaxValue;
+
+  // initialize the input data
+  for (size_t i = 0; i < kSize; i++)
+    a[i] = rand() % kMaxValue;
+
+#if defined(FPGA_EMULATOR)
+  INTEL::fpga_emulator_selector selector;
+#else
+  INTEL::fpga_selector selector;
+#endif
+
+  // Run the kernel with different values of the private_copies
+  // attribute, to determine the optimal concurrency.
+  // In this case, the optimal private_copies is <TODO: Need to put a number
+  // here> since this achieves the highest GFlops. Higher values of
+  // private_copies consume additional RAM without increasing GFlops.
+  SimpleMathWithShift<0>(selector, a, shift, R0);
+  SimpleMathWithShift<1>(selector, a, shift, R1);
+  SimpleMathWithShift<2>(selector, a, shift, R2);
+  SimpleMathWithShift<3>(selector, a, shift, R3);
+  SimpleMathWithShift<4>(selector, a, shift, R4);
+  SimpleMathWithShift<5>(selector, a, shift, R5);
+
+  // compute the actual result here
+  int gr = GoldenResult(a, shift);
+
+  // verify the results are correct
+  if (gr != R0[0]) {
+    std::cout << "private_copies 0: mismatch: " << R0[0] << " != " << gr
+              << " (kernel != expected)" << '\n';
+    success = false;
+  }
+
+  if (gr != R1[0]) {

```

Comment:
You can do this check in a function that you call for each array
This will remove duplicated code :)